File: wiLodPortal/app/controllers/Application.java
package controllers;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;

import play.api.templates.Html;
import play.mvc.Controller;
import play.mvc.Result;
import views.html.index;
import views.html.chart;

import com.google.gson.Gson;
import com.thoughtworks.xstream.XStream;
import com.thoughtworks.xstream.io.xml.DomDriver;

import es.weso.business.ComponentManagement;
import es.weso.business.CountryGroupManagement;
import es.weso.business.CountryManagement;
import es.weso.business.DatasetManagement;
import es.weso.business.IndicatorManagement;
import es.weso.business.ObservationManagement;
import es.weso.business.SubindexManagement;
import es.weso.business.WeightSchemaManagement;
import es.weso.business.impl.ComponentManager;
import es.weso.business.impl.CountryGroupManager;
import es.weso.business.impl.CountryManager;
import es.weso.business.impl.DatasetManager;
import es.weso.business.impl.IndicatorManager;
import es.weso.business.impl.ObservationManager;
import es.weso.business.impl.SubindexManager;
import es.weso.business.impl.WeightSchemaManager;
import es.weso.wirouter.CountryRouteParser;
import es.weso.wirouter.YearRouteParser;
import es.weso.wirouter.country.CountryExpr;
import es.weso.wirouter.year.YearExpr;

public class Application extends Controller {

	private static Gson gson = new Gson();
	private static XStream xstream = new XStream(new DomDriver());
	private static ComponentManagement componentManager = ComponentManager
			.getInstance();
	private static CountryGroupManagement countryGroupManager = CountryGroupManager
			.getInstance();
	private static CountryManagement countryManager = CountryManager
			.getInstance();
	private static DatasetManagement datasetManager = DatasetManager
			.getInstance();
	private static IndicatorManagement indicatorManager = IndicatorManager
			.getInstance();
	private static ObservationManagement observationManager = ObservationManager
			.getInstance();
	private static SubindexManagement subindexManager = SubindexManager
			.getInstance();
	private static WeightSchemaManagement weightSchemaManager = WeightSchemaManager
			.getInstance();

	private static Result render(Object obj, String format) {
		if (request().accepts("text/html")) {
			return ok(chart.render("Chart Generator", obj));
		}
		if (request().accepts("text/turtle")) {
			return renderTTL();
		}
		if (request().accepts("application/rdf+xml")) {
			return renderRDF();
		}
		if (request().accepts("application/json")) {
			return renderJSON(obj);
		}
		if (request().accepts("application/xml")) {
			return renderXML(obj);
		}
		return badRequest(); // Should return 415
	}

	private static Result renderXML(Object obj) {
		return ok(xstream.toXML(obj));
	}

	private static Result renderJSON(Object obj) {
		return ok(gson.toJson(obj));
	}

	private static Result renderTTL() {
		try {
			return renderLD("text");
		} catch (Exception e) {
			return internalServerError();
		}
	}

	private static Result renderRDF() {
		try {
			return renderLD("xml");
		} catch (Exception e) {
			return internalServerError();
		}
	}

	private static Result renderLD(String output) throws Exception {
		// FIXME This is fuseki dependant and dirty...
		System.out.println(request().path());
		String relativeUri = URLEncoder.encode(
				request().path().substring(request().path().indexOf('/', 0)),
				"UTF-8");
		String url = "http://localhost:3030/computex/query?query=CONSTRUCT%7B%3Chttp%3A%2F%2Fdata.webfoundation.org%2Fwebindex%2Fv2013"
				+ relativeUri
				+ "%3E+%3Fp+%3Fl%7DWHERE%7B%3Chttp%3A%2F%2Fdata.webfoundation.org%2Fwebindex%2Fv2013"
				+ relativeUri + "%3E+%3Fp+%3Fl%7D&output=" + output;
		URL obj = new URL(url);
		HttpURLConnection con = (HttpURLConnection) obj.openConnection();
		BufferedReader in = new BufferedReader(new InputStreamReader(
				con.getInputStream()));
		String inputLine;
		ByteArrayOutputStream baos = new ByteArrayOutputStream();
		while ((inputLine = in.readLine()) != null) {
			baos.write(inputLine.getBytes());
			baos.write("\n".getBytes());
		}
		in.close();
		return ok(baos.toString());
	}

	public static Result index() {
		return ok(index.render("Your new application is ready."));
	}

	public static Result components() {
		return render(componentManager.getAll(), null);
	}

	public static Result component(String id) {
		return render(componentManager.getOne(id), null);
	}

	public static Result countries() {
		return render(countryManager.getAllCountries(), null);
	}

	public static Result country(String code) {
		return render(countryManager.getCountry(code), null);
	}

	public static Result datasets() {
		return render(datasetManager.getAll(), null);
	}

	public static Result dataset(String id) {
		return render(datasetManager.getOne(id), null);
	}

	public static Result indicators() {
		return render(indicatorManager.getAllIndicators(), null);
	}

	public static Result indicator(String id) {
		return render(indicatorManager.getIndicatorByURI(id), null);
	}

	public static Result observations() {
		return render(observationManager.getAllObservations(), null);
	}

	public static Result observation(String id) {
		return render(observationManager.getObservationByURI(id), null);
	}

	public static Result observationByCountry(String country, int year,
			String indicator) {
		// TODO Redirect
		return render(
				observationManager.getAllObservationsByCountries(
						Collections.singleton(country),
						Collections.singleton(indicator),
						Collections.singleton(year)), null);
	}

	public static Result regions() {
		return render(countryGroupManager.getAllCountryGroups(), null);
	}

	public static Result region(String name) {
		return render(countryGroupManager.getCountryGroup(name), null);
	}

	public static Result subindexes() {
		return render(subindexManager.getAll(), null);
	}

	public static Result subindex(String id) {
		return render(subindexManager.getOne(id), null);
	}

	public static Result weightSchemas() {
		return render(weightSchemaManager.getAll(), null);
	}

	public static Result weightSchema(String id) {
		return render(weightSchemaManager.getOne(id), null);
	}

	public static Result compare(String countries, String years,
			String indicators) {
		return render(observationManager.getAllObservationsByCountries(
				parseCountryCodes(countries), parseObservations(indicators),
				parseYears(years)), null);
	}

	/**
	 * Parses a expression to get the country codes contained in that expression
	 * 
	 * @param expression
	 *            The expression to be parsed
	 * @return The country codes contained in that expression
	 */
	private static Collection<String> parseCountryCodes(String expression) {
		Collection<String> countries = new HashSet<String>();
		Collection<CountryExpr> exprs = new CountryRouteParser()
				.parseRoute(expression);
		for (CountryExpr expr : exprs) {
			countries.addAll(expr.getCountryCodes());
		}
		return countries;
	}

	/**
	 * Parses a expression to get the observations contained in that expression
	 * 
	 * @param expression
	 *            The expression to be parsed
	 * @return The observations contained in that expression
	 */
	private static Collection<String> parseObservations(String expression) {
		String[] arr = expression.split(",");
		Collection<String> observations = new HashSet<String>(arr.length);
		for (String expr : arr) {
			observations.add(expr.trim());
		}
		return observations;
	}

	/**
	 * Parses a expression to get the years contained in that expression
	 * 
	 * @param expression
	 *            The expression to be parsed
	 * @return The years contained in that expression
	 */
	private static Collection<Integer> parseYears(String expression) {
		Collection<Integer> years = new HashSet<Integer>();
		Collection<YearExpr> exprs = new YearRouteParser()
				.parseRoute(expression);
		for (YearExpr expr : exprs) {
			years.addAll(expr.getYears());
		}
		return years;
	}

}

File: wiLodPortal/app/controllers/Application.scala
package controllers

import java.util.Collection
import java.util.Collections

import scala.collection.JavaConversions.asScalaBuffer
import scala.concurrent.Await
import scala.concurrent.duration._

import es.weso.wiLodPortal.business.impl.ComponentManager
import es.weso.wiLodPortal.business.impl.CountryGroupManager
import es.weso.wiLodPortal.business.impl.CountryManager
import es.weso.wiLodPortal.business.impl.DatasetManager
import es.weso.wiLodPortal.business.impl.IndicatorManager
import es.weso.wiLodPortal.business.impl.ObservationManager
import es.weso.wiLodPortal.business.impl.SubindexManager
import es.weso.wiLodPortal.business.impl.WeightSchemaManager
import es.weso.wirouter.CountryRouteParser
import es.weso.wirouter.YearRouteParser
import es.weso.wirouter.year.YearExpr
import play.api.libs.concurrent.Execution.Implicits.defaultContext
import play.api.libs.ws.WS
import play.api.mvc.Action
import play.api.mvc.Controller

object Application extends Controller with Renders {

  val componentManager = ComponentManager.getInstance
  val countryGroupManager = CountryGroupManager.getInstance
  val countryManager = CountryManager.getInstance
  val datasetManager = DatasetManager.getInstance
  val indicatorManager = IndicatorManager.getInstance
  val observationManager = ObservationManager.getInstance
  val subindexManager = SubindexManager.getInstance
  val weightSchemaManager = WeightSchemaManager.getInstance

  val url = "http://localhost:8080/pubby/"
  val Data = "data"
  val Page = "page"

  def index = Action { implicit request =>
    Ok(views.html.index("Your new application is ready."))
  }

  def components = Action { implicit request =>
    render(componentManager.getAll(), null)
  }

  def component(id: String) = Action { implicit request =>
    render(componentManager.getOne(id), null)
  }

  def countries() = Action { implicit request =>
    render(countryManager.getAllCountries(), null)
  }

  def country(code: String) = Action { implicit request =>
    render(countryManager.getCountry(code), null)
  }

  def datasets() = Action { implicit request =>
    render(datasetManager.getAll(), null)
  }

  def dataset(id: String) = Action { implicit request =>
    render(datasetManager.getOne(id), null)
  }

  def indicators() = Action { implicit request =>
    render(indicatorManager.getAllIndicators(), null);
  }

  def indicator(id: String) = Action { implicit request =>
    render(indicatorManager.getIndicatorByURI(id), null);
  }

  def observations() = Action { implicit request =>
    render(observationManager.getAllObservations(), null)
  }

  def observation(id: String) = Action { implicit request =>
    render(observationManager.getObservationByURI(id), null)
  }

  def observationByCountry(country: String, year: Int, indicator: String) = Action { implicit request =>
    // TODO Redirect
    render(
      observationManager.getAllObservationsByCountries(
        Collections.singleton(country),
        Collections.singleton(indicator),
        Collections.singleton(year)), null)
  }

  def regions() = Action { implicit request =>
    render(countryGroupManager.getAllCountryGroups(), null)
  }

  def region(name: String) = Action { implicit request =>
    render(countryGroupManager.getCountryGroup(name), null)
  }

  def subindexes() = Action { implicit request =>
    render(subindexManager.getAll(), null)
  }

  def subindex(id: String) = Action { implicit request =>
    render(subindexManager.getOne(id), null)
  }

  def weightSchemas() = Action { implicit request =>
    render(weightSchemaManager.getAll(), null)
  }

  def weightSchema(id: String) = Action { implicit request =>
    render(weightSchemaManager.getOne(id), null)
  }

  def compare(countries: String, years: String, indicators: String) = Action { implicit request =>
    render(observationManager.getAllObservationsByCountries(
      parseCountryCodes(countries), parseObservations(indicators),
      parseYears(years)), null)
  }

  def fallback(path: String) = Action { implicit request =>

    def inner(mimeType: String, mode: String, suffix: Option[String] = None) = {
      val query = new StringBuilder(url).append(mode).append("/").append(path)
      suffix match {
        case Some(suffix) => query.append(suffix)
        case None =>
      }
      println(query.toString)
      val future = WS.url(query.toString).withHeaders("Accept" -> mimeType.toString()).get().map {
        response => Ok("" + response.body).as(mimeType)
      }
      Await.result(future, 60 seconds)
    }

    render {
      case Html() => inner(Html.mimeType, Page)
      case Turtle() => inner(Turtle.mimeType, Data, Some("?output=ttl"))
      case XTurtle() => inner(Turtle.mimeType, Data, Some("?output=ttl"))
      case RdfXML() => inner(RdfXML.mimeType, Data, Some("?output=xml"))
      case Json() => inner(Json.mimeType, Data)
      case Xml() => inner(RdfXML.mimeType, Data, Some("?output=xml"))
      case _ => BadRequest
    }

  }

  /**
   * Parses a expression to get the country codes contained in that expression
   *
   * @param expression
   *            The expression to be parsed
   * @return The country codes contained in that expression
   */
  private def parseCountryCodes(expression: String): Collection[String] = {
    val countries = new java.util.HashSet[String]();
    val exprs = new CountryRouteParser().parseRoute(expression);
    for (expr <- exprs) {
      countries.addAll(expr.getCountryCodes());
    }
    return countries;
  }

  /**
   * Parses a expression to get the observations contained in that expression
   *
   * @param expression
   *            The expression to be parsed
   * @return The observations contained in that expression
   */
  private def parseObservations(expression: String): Collection[String] = {
    val arr = expression.split(",")
    val observations = new java.util.HashSet[String](arr.length)
    for (expr <- arr) {
      observations.add(expr.trim)
    }
    observations;
  }

  /**
   * Parses a expression to get the years contained in that expression
   *
   * @param expression
   *            The expression to be parsed
   * @return The years contained in that expression
   */
  private def parseYears(expression: String): Collection[Integer] = {
    val years = new java.util.HashSet[Integer]()
    val exprs: List[YearExpr] = new YearRouteParser().parseRoute(expression).toList
    for (expr <- exprs) {
      years.addAll(expr.getYears())
    }
    years
  }
}